File: src/test/java/com/giwankim/leetcode/ReorderLogFilesTest.java
package com.giwankim.leetcode;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class ReorderLogFilesTest {
  @ParameterizedTest
  @MethodSource
  void reorderLogFiles(String[] logs, String[] expected) {
    String[] actual = new ReorderLogFiles().reorderLogFiles(logs);
    assertThat(actual).isEqualTo(expected);
  }

  private static Stream<Arguments> reorderLogFiles() {
    return Stream.of(
        Arguments.of(
            new String[] {
              "dig1 8 1 5 1", "let1 art can", "dig2 3 6", "let2 own kit dig", "let3 art zero"
            },
            new String[] {
              "let1 art can", "let3 art zero", "let2 own kit dig", "dig1 8 1 5 1", "dig2 3 6"
            }),
        Arguments.of(
            new String[] {"a1 9 2 3 1", "g1 act car", "zo4 4 7", "ab1 off key dog", "a8 act zoo"},
            new String[] {"g1 act car", "a8 act zoo", "ab1 off key dog", "a1 9 2 3 1", "zo4 4 7"}));
  }
}

File: leetcode-kt/src/test/kotlin/com/giwankim/leetcode/ReorderLogFilesTest.kt
package com.giwankim.leetcode

import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withTests
import io.kotest.matchers.shouldBe

class ReorderLogFilesTest :
    FunSpec({
        val sut = ReorderLogFiles()

        context("reorder log files") {
            withTests(
                nameFn = { (logs, expected) ->
                    "logs=${logs.contentToString()}, expected=${expected.contentToString()}"
                },
                ReorderLogFilesCase(
                    arrayOf("dig1 8 1 5 1", "let1 art can", "dig2 3 6", "let2 own kit dig", "let3 art zero"),
                    arrayOf("let1 art can", "let3 art zero", "let2 own kit dig", "dig1 8 1 5 1", "dig2 3 6"),
                ),
                ReorderLogFilesCase(
                    arrayOf("a1 9 2 3 1", "g1 act car", "zo4 4 7", "ab1 off key dog", "a8 act zoo"),
                    arrayOf("g1 act car", "a8 act zoo", "ab1 off key dog", "a1 9 2 3 1", "zo4 4 7"),
                ),
            ) { (logs, expected) ->
                sut.reorderLogFiles(logs) shouldBe expected
            }
        }
    })

private data class ReorderLogFilesCase(
    val logs: Array<String>,
    val expected: Array<String>,
)
